In [ ]:
# Assignment 2

This assignment will be concerned with demonstrating the following course contents:
* Relational databases (SQL) and MongoDB (NoSQL)
* Graph Databases
* Streaming algorithms



## SQL and NoSQL

SQL databases has over the last decades been the core norm of persisting data into non-volatile memory. Although SQL is a great way to represent certain data sets, it can be quite time expensive when queries starts to compute a high number of large cartesian products between tables. To mitigate some of these performance problems, there has been an increasing interest within the area of NoSQL (i.e. a database that is not meant to have normalized relations). MongoDB is a popular NoSQL database and provides a schema-less persistency of the data in a json-like format (called bson). In contrary to SQL, the data in a NoSQL database should be denormalized as much as possible (that is, you should store it wihin a nested structure). This makes MongoDB especially interesting when storing large documents or highly detailed log-files with a high throughput.

In this exercise we're going to use the Northwind database. To give the reader a idea of how this database looks like, from a relational perspective, a ER diagram is displayed below.

<img src="http://i3.codeplex.com/download?ProjectName=northwinddatabase&DownloadId=269240" width="800" height="800" />

The equivalent MongoDB database, that we are going to use, has onfortunately not been denormalized into one collection. Because of this, the way MongoDB is being used in this exercise is somehow _a special case_.

### Exercise 1

_Task: Establish connection to MongoDB and SQLite_

In the first exercise, we are going to make a basic connection to both SQLite (a light-weight SQL engine that persists data into files) and to MongoDB.

#### Connection to SQLite

The first thing we need to do, is to import sqlite3, which is the Python module we're going to use.

In [1]:
import sqlite3 as lite

We can create a small function that returns a connection to the northwind.db (a file in the same directory).

In [2]:
def get_sqlite_connection():
    conn = lite.connect('northwind.db')
    conn.text_factory = str
    return conn

From here on, getting a connection to the SQLite DB is easy...

In [ ]:
conn = get_sqlite_connection()

Given the `conn` object we can create a cursor, which is used to query the database.

In [ ]:
cur = conn.cursor()

Using the cursor `cur` we can can retrieve all the products by making the sql query `SELECT * FROM PRODUCTS`.

In [77]:
cur.execute('SELECT * FROM PRODUCTS')

The query has been executed and the result set can be obtained by calling fetchall. Here, we are just going to see the first element.

In [78]:
cur.fetchall()[:1] # Take the first element

[(1, 'Chai', 1, 1, '10 boxes x 20 bags', 18, 39, 0, 10, '0')]

We can query the database once again, this time, however, we are going to sort the result set using the `customerid`, which is a column in the table `CUSTOMERS`.

In [79]:
cur.execute('SELECT * FROM customers ORDER BY customerid ASC')
cur.fetchall()[:1]

[('ALFKI',
  'Alfreds Futterkiste',
  'Maria Anders',
  'Sales Representative',
  'Obere Str. 57',
  'Berlin',
  None,
  '12209',
  'Germany',
  '030-0074321',
  '030-0076545')]

#### Connection to MongoDB

Connecting to MongoDB is slightly more difficult, than connecting to a SQLite DB, since MongoDB is a client-server database engine and has to be configured.

First, we need to install MongoDB. After MongoDB is installed, we need to run a shell script that imports the northwind database into MongoDB. After this, we can start the MongoDB service on localhost and the default port using `mongod` in bash.

<img src="img/mongod-service-start.png" width="800" height="800" />

After we've done this, we can `pip install pymongo`, and we are finally ready to start interacting with Northwind MongoDB through Python.

In [5]:
from pymongo import MongoClient
import pymongo

We can create a function, that connects to MongoDB and then returns a connection to "Nortwind"

In [6]:
def get_mongo_connection():
    client=MongoClient('localhost',27017)
    return client['Northwind']   # Get the database

At this point, we can easily make return an MongoDB conn instance

In [7]:
c = get_mongo_connection()

We have established a connection and can use it to retrieve all the products and list the first entry.

In [83]:
products = c.products.find()
list(products)[0]

{u'CategoryID': 1,
 u'Discontinued': 0,
 u'ProductID': 1,
 u'ProductName': u'Chai',
 u'QuantityPerUnit': u'10 boxes x 20 bags',
 u'ReorderLevel': 10,
 u'SupplierID': 1,
 u'UnitPrice': 18.0,
 u'UnitsInStock': 39,
 u'UnitsOnOrder': 0,
 u'_id': ObjectId('5609b12899ded9537b06c7e5')}

Although, this is just a dict, it can be easily seen as a bson object, which is the original way that MongoDB persists data.

If we wanted to make yet another query, but this time we wanted to sort the result, we can apply a sort filter to a specific bson key. From here on, we're going to use Pandas DataFrame since they provide an easy way of pretty-printing tabular data.

In [84]:
c = get_mongo_connection()
customers = c.customers.find().sort('CustomerID', pymongo.ASCENDING)
top10 = [customer['CustomerID'] for customer in customers][:10]
pd.DataFrame(top10, columns=['CustomerID'])

CustomerID
0      ALFKI
1      ANATR
2      ANTON
3      AROUT
4      BERGS
5      BLAUS
6      BLONP
7      BOLID
8      BONAP
9      BOTTM

The list above shows the top 10 customer ids, sorted in ascending order, from the customer collection.

### Exercise 2

In [ ]:
cur.execute("SELECT * FROM customers AS c INNER JOIN orders AS o ON c.customerid=o.customerid where o.customerid='ALFKI'")
pd.DataFrame(cur.fetchall())

In [ ]:
for customer in c.customers.find({'CustomerID':'ALFKI'}):
    for order in c.orders.find({'CustomerID': customer['CustomerID']}):
        for details in c['order-details'].find({'OrderID': order['OrderID']}):
            joined_result = order.copy()
            joined_result.update(details)
r = cur.fetchall()

# Pretty print using pandas
pd.DataFrame([joined_result.values()], columns=joined_result.keys())

### Exercise 3
_Get all orders (with products) made by ALFKI that contain at least 2 products._

In [15]:
import pandas as pd

In [ ]:
cur.execute('SELECT * FROM customers AS c INNER JOIN orders AS o ON c.customerid=o.customerid '
                    ' INNER JOIN "Order Details" AS od ON od.orderid=o.orderid '
                    ' INNER JOIN products AS p ON p.productid=od.productid '
                    " WHERE c.customerid='ALFKI'"
                    ' GROUP BY o.orderid HAVING count(distinct od.productid) >= 2')
r = cur.fetchall()
pd.DataFrame(r)

In [ ]:
from bson.code import Code
c = get_mongo_connection()
reducer = Code(
        """
            function(obj, prev){
              prev.count++;
            }
        """)
orders = []
for customer in c.customers.find({'CustomerID':'ALFKI'}):
    for order in c.orders.find({'CustomerID': customer['CustomerID']}):
        results = c['order-details'].group(key={"OrderID": 1}, condition={}, initial={"count": 0}, reduce=reducer)
        for details in results:
            if details['OrderID'] == order['OrderID'] and details['count'] >= 2:
                orders.append(order)
                joined_result = order.copy()
                joined_result.update(details)
                
# Pretty print using pandas
pd.DataFrame(orders)

#### Exercise 4

In [ ]:
class TestExercise54(unittest.TestCase):
    def test_sqlite_count_products(self):
        conn = connect_sqlite()
        cur = conn.cursor()
        cur.execute('SELECT count(productid) FROM "Order Details" where productid=7')
        print 'SQLite Count: ' + str(cur.fetchone()[0])

    def test_mongo_count_products(self):
        c = get_mongo_connection()
        print 'MongoDB Count: ' + str(c['order-details'].find({'ProductID': 7}).count())

#### Exercise 5

In [74]:
stmt = """
SELECT count(p.productid), p.productname FROM orders
INNER JOIN [Order Details] AS od ON od.orderid=orders.orderid
INNER JOIN products AS p ON p.productid=od.productid
WHERE CustomerID IN
            (SELECT CustomerID FROM orders
             INNER JOIN [Order Details] ON [Order Details].orderid=orders.orderid
             WHERE [Order Details].productid=7)
GROUP BY od.productid
"""
cur.execute(stmt)
#print list(cur.fetchall())
df_sql = pd.DataFrame(cur.fetchall(), columns=['count','ProductName'])
df_sql

count                       ProductName
0       8                              Chai
1      18                             Chang
2       7                     Aniseed Syrup
3       6      Chef Anton's Cajun Seasoning
4       6            Chef Anton's Gumbo Mix
5       5      Grandma's Boysenberry Spread
6      29   Uncle Bob's Organic Dried Pears
7       5        Northwoods Cranberry Sauce
8       1                   Mishi Kobe Niku
9      16                             Ikura
10     12                    Queso Cabrales
11      7         Queso Manchego La Pastora
12     15                             Konbu
13      7                              Tofu
14      2                      Genen Shouyu
15     14                           Pavlova
16     16                      Alice Mutton
17      9                  Carnarvon Tigers
18      9        Teatime Chocolate Biscuits
19      7            Sir Rodney's Marmalade
20      9               Sir Rodney's Scones
21      6               Gustaf's Kn�ckebr�d
22      8                          Tunnbr�d
23     15                Guaran� Fant�stica
24      5           NuNuCa Nu�-Nougat-Creme
25     10               Gumb�r Gummib�rchen
26      4                Schoggi Schokolade
27      8                 R�ssle Sauerkraut
28     10           Th�ringer Rostbratwurst
29     10             Nord-Ost Matjeshering
..    ...                               ...
46     10                     Zaanse koeken
47      2                         Chocolade
48      7                          Maxilaku
49      2                  Valkoinen suklaa
50     16             Manjimup Dried Apples
51      7                          Filo Mix
52      8                     Perth Pasties
53     13                         Tourti�re
54     12                      P�t� chinois
55     18            Gnocchi di nonna Alice
56      8                    Ravioli Angelo
57      4            Escargots de Bourgogne
58     19              Raclette Courdavault
59     19                 Camembert Pierrot
60      7                    Sirop d'�rable
61     22                    Tarte au sucre
62      7                      Vegie-spread
63     13         Wimmers gute Semmelkn�del
64     12  Louisiana Fiery Hot Pepper Sauce
65      4         Louisiana Hot Spiced Okra
66     14               Scottish Longbreads
67      9                  Gudbrandsdalsost
68     13                     Outback Lager
69     16                       Flotemysost
70     16            Mozzarella di Giovanni
71      4                        R�d Kaviar
72      5                     Longlife Tofu
73     17              Rh�nbr�u Klosterbier
74     12                      Lakkalik��ri
75     15   Original Frankfurter gr�ne So�e

[76 rows x 2 columns]

c = get_mongo_connection()
from bson.code import Code
reducer = Code(
        """
            function(obj, prev){
              prev.count++;
            }
        """)
#results = c['order-details'].group(key={"ProductID": 1, 'CustomerID':2}, condition={}, initial={"count": 0}, reduce=reducer)
#for x in results:
#    print x
customers = {}
products = {}
for details in c['order-details'].find({'ProductID': 7}):
    for order in c.orders.find({'OrderID':details['OrderID']}):
        customers[order['CustomerID']]= 0

for customer_id in customers.keys():
    for r in c.orders.find():
        if r['CustomerID'] == customer_id:
            for details in c['order-details'].find({'OrderID':r['OrderID']}):

                product_name = c.products.find({'ProductID': details['ProductID']})[0]['ProductName']

                if product_name in products.keys():
                    products[product_name] += 1
                else:
                    products[product_name] = 1

df_mongo = pd.DataFrame(products.keys(), columns=['ProductName']).sort('ProductName')
df_mongo

## Streaming Algorithms

We use streaming algorithms when the given data is of such a size that it can not be stored at all, and where we only get to observe each sample once. Also, Wikipedia adds that *These algorithms have limited memory available to them (much less than the input size) and also limited processing time per item. These constraints may mean that an algorithm produces an approximate answer based on a summary or "sketch" of the data stream in memory.*

Three of these kind of algorithms will be examined here.

### Exercise 7.1 - Bloom filter

The Bloom filter is a probabilistic data structure, that is used to test whether an element is a member of a set. Only false positive matches can occur (the element is shown as a member though it is not - "possibly in set", or "definitely not in the set"), therefore it is said that a Bloom filter has a 100% recall rate. 

In this exercise, a such filter is to be built, which is able to check the words from a text written by Shakespeare against a dictionary representing all words of the English language (*shakespeare.txt*, *dict*).

#### Implement a Bloom Filter as a Python class

In [199]:
import numpy

class bloom_filter:
    
    # ctor of the filter
    def __init__(self, bm, hashf, hashfk):
        # initializing the bit array to zeroes
        self.__bitarray = numpy.zeros(bm)
        self.m = bm
        
        # taking the hash function. it should take two parameters:
        # the string to be hashed, and the parameter of the hash, which will be
        # just the no. of it now
        self.__hash = hashf
        
        # storing the number of hash functions
        self.__k = hashfk
        
    def __get_bitz(self, item):
        # calls the hash function the given "k" times, returns the hash values in a list
        # it is divided by % in order to limit the results to the 'm' range
        return [self.__hash(item, kc) % self.m for kc in range(self.__k)]
    
    def add(self, item):
        try:
            numpy.put(self.__bitarray, self.__get_bitz(item), 1)
        except IndexError:
            print 'That offset isnt here!'
        
    def lookup(self, item):
        # get the set bit (indexes) for the given item
        bitz = self.__get_bitz(item)
        
        # check it against the stored bitstring
        # returns False if not all of the items are nonzero
        return numpy.count_nonzero(numpy.take(self.__bitarray, bitz)) == len(bitz)        

The Bloom filter class shown above is quite general - due to the fact, that many of the parameters are up to the user.

The desired size of the filter (= hash table size) (*m*), the hash function and the number of hash functions (*k*) are specified when instantiating the class. According to the exercise quite, this time we will use a bitstring with 1000000 (one-million) bits, and the number of the hash functions are calculated as: $m/n * ln(2)$. The *n* in the equation is the expected number of distinct elements in the input (= set size), equals with the length of our English dictionary in lines.

In [200]:
import math # due to log() and ceil()

m = 1000000
n = sum(1 for line in open('dict'))

k = int(math.ceil(float(m) / n  * math.log(2)))

print 'The exact values then: m = %d, n = %d, k = %d' % (m, n, k)

The exact values then: m = 1000000, n = 235886, k = 3


The hash function is the **MurmurHash3** this time, which is just passed to the class as a lambda function. Its second parameter (the index of the given hash function in the Bloom filter) will be used as a seed to the hash function, yielding a different (albeit of course reproduceable) hash for the same input string.

In [201]:
import mmh3

# using the MurmurHash3, and the number of the hash function as the seed
hash_function = lambda x, y : mmh3.hash(x) if y == None else mmh3.hash(x, y)

It has to be noted there, that we can use a cryptographically unsafe hash function for the Bloom-filter, we only care about an uniform distribution of the filter output values.

Then, the class can be created finally:

In [202]:
filter = bloom_filter(m, hash_function, k)

The Shakespeare-text was processed using the following Unix command:

`tr -s  '[:blank:]\n' '\n' < shakespeare.txt | tr 'A-Z' 'a-z' | tr -cd [:lower:][:cntrl:] | sed '/^$/d' | sort | uniq > shakespeare_words.txt`

in order to get a file containing only the separate, unique words.

Then, the Bloom filter is filled with all the words from *dict*:

In [203]:
for line in open('dict'):
    filter.add(line.strip())

In [204]:
def bloom(bloomfilter, filename):
    not_in_filter = []
    for word in open(filename):
        if not filter.lookup(word.strip()):
            not_in_filter.append(word.strip())
    return not_in_filter

In [205]:
not_in_filter = bloom(filter, 'shakespeare_words.txt')

In [206]:
print 'There are', len(not_in_filter), 'words in the Shakespeare text, which are not in the Bloom filter.'

There are 800 words in the Shakespeare text, which are not in the Bloom filter.


In order to jump to some conclusions, we chose some examples from this *not_in_filter* list above.

Firstly, there are the words in the list, which are obviously spelt wrongly:
- *abandond, accustomd, addressd, breakpromise, complaind, femalewhich* etc.

But, also there are words present, which look right:
- *abused, bled, egypt, feet*

And names, which are not necessarily present in a dictionary of english words.
- *adam, celia, charles* etc.

Worth noting too, that the list also apparently contains many words in the plural form (ending with an *-s*) or the third person singular with the same *-s* ending. Obviously then, a proper spell checker implementation would work with the stems (roots??) of the words - but that is outside of the scope of this exercise.

Now, the approach with the Bloom filter will be compared to a simpler approach, which just compares the words of the dictionary to each of the distinct words of the Shakespeare text.

In [207]:
def file_strip_dict(filename):
    file_contents = []
    for line in open(filename):
        file_contents.append(line.strip())
    return file_contents

# storing all the words in a list
english_dict = file_strip_dict('dict')

# storing all uniques Shakespeare-words in another list
shakespeare_list = file_strip_dict('shakespeare_words.txt')

And then, the even simpler spellchecker:

In [208]:
def simple_spellcheck(input_words, dictionary):
    not_found = []
    for input_word in input_words:
        if input_word not in dictionary:
            not_found.append(input_word)
    return not_found

In [209]:
not_in_filter_2 = simple_spellcheck(shakespeare_list, english_dict)

In [210]:
print 'There are', len(not_in_filter_2), 'words in the Shakespeare text, which are not in the dictionary.'

There are 928 words in the Shakespeare text, which are not in the dictionary.


More words are found to be spelt wrongly using this method, which might look surprising, as we expected the Bloom filter to be "worse" due to using probabilities. However, it should be remembered again: the Bloom filter's answers mean something is **possibly in** the set, or **definitely not** in the set. Therefore, it could show that some wrongly spelt words are in the dictionary (meaning it passes them as correctly spelt), though they arent't there.
Consequently, having less results here mean a worse performance.

To further elaborate on the spell-checker, let's test briefly the "usual" approach for Bloom-filters too: checking the incoming word first against it, and in the case of a **possibly in** answer, looking it up in the dictionary, in order to verify whether the word really does exist in the set.

In [211]:
def usual_bloom(bloomfilter, dictionary, words):
    actually_not = []
    not_in_bloom = []
    for word in words:
        if bloomfilter.lookup(word):
            # checking against the dictionary
            if word not in dictionary:
                # it wasn't in the dictionary, so the Bloom filter misled us
                actually_not.append(word)
        else:
            not_in_bloom.append(word)
    
    return actually_not, not_in_bloom              

In [212]:
bloom_miss, bloom_not = usual_bloom(filter, english_dict, shakespeare_list)

In [213]:
print len(bloom_miss), 'words were not in the dictionary, even though the Bloom filter said they might be.'
print len(bloom_not), 'words were not present in the Bloom filter.'
print 'In total,', len(bloom_miss) + len(bloom_not), 'words of the Shakespeare text were not found in the dictionary.'

128 words were not in the dictionary, even though the Bloom filter said they might be.
800 words were not present in the Bloom filter.
In total, 928 words of the Shakespeare text were not found in the dictionary.


The method worked - and gave the same number of words as a result like with comparing the two files word-by-word.

Thus, it is proven, that the Bloom filter gives *false positive* results - can be true for items, which are actually not a part of the set.

#### What is the speed difference?
Let's take a look at the timings of the three different methods demonstrated above! The *%timeit* option of iPython was used, and while the exact time values are only representative to the computer this exercise was made on, their magnitude will be approximately the same on other systems too.

- *simple_spellcheck*: 1 loops, best of 3: 6.25 s per loop
- *usual_bloom*: 1 loops, best of 3: 4.58 s per loop
- *bloom* 100 loops, best of 3: 9.95 ms per loop

It can be concluded, that compared to the methods using lookups from a complete dictionary, the plain Bloom filter approach is almost a magnitude faster. It has the drawback of false positives on the other hand.
The two other approaches are much slower, although it is worth noting that they are not optimized here.

#### Which words are reported to be in the dictionary but are actually not in there?

In [214]:
bloom_miss

['abandonwhich',
 'ages',
 'alls',
 'april',
 'atomies',
 'attending',
 'audreyas',
 'augmenting',
 'bands',
 'basest',
 'bells',
 'berhymed',
 'boldend',
 'brawls',
 'breaks',
 'brotherno',
 'brows',
 'burghers',
 'burthen',
 'cannons',
 'coroners',
 'cries',
 'crookedpated',
 'crowns',
 'cupid',
 'curvets',
 'demonstrating',
 'dennis',
 'depends',
 'diseases',
 'dominions',
 'earthquakes',
 'england',
 'entreaties',
 'everas',
 'evils',
 'exits',
 'extremest',
 'fathers',
 'features',
 'fells',
 'fields',
 'filld',
 'flattered',
 'followed',
 'forestborn',
 'france',
 'ganymede',
 'gets',
 'gipsies',
 'goats',
 'graces',
 'grecian',
 'hadst',
 'hands',
 'hanged',
 'having',
 'hawthorns',
 'heads',
 'heres',
 'horses',
 'humbled',
 'husbands',
 'inclined',
 'joves',
 'leavethe',
 'lets',
 'liest',
 'liquors',
 'looked',
 'lookest',
 'loved',
 'lungs',
 'maidens',
 'mans',
 'mingled',
 'misprised',
 'mossd',
 'newscrammed',
 'oerrun',
 'offenders',
 'overcame',
 'overthrown',
 'pains',

### Exercise 7.2 - Flajolet-Martin algorithm

The *Flajolet-Martin* algorithm is an algorithm for approximating the number of distinct elements in a stream with a single pass and space-consumption which is logarithmic in the maximum number of possible distinct elements in the stream. 

In-pseudo-code, that is:

```init BITMAP[] = 0
for X in M:
    index = r(hash(x)) // where R is the index of the first 1 seen from the LSB in the binary form
    BITMAP[index] = 1```

Then, the approximate number of unique values in *M* is:

$2^{R} / 0.77351$

where *R* is the last index (again, from LSB), where ```BITMAP[index] = 0```.

The algorithm was made as a Python class:

In [190]:
from bitstring import BitArray

class flajolet_martin:
    
    def __init__(self, hashf, hashparam=None):
        # size fixed to 32 bits
        self.__bitarray_size = 32
        
        # initializing the bit array to zeroes
        self.__bitarray = BitArray(self.__bitarray_size) # thus it is from 0 - 2**basize
        
        # taking the hash function. it should take two parameters:
        # the string to be hashed, and the parameter of the hash, which will be
        # just the no. of it now
        self.__hash = hashf
        
        self.__hashp = hashparam
    
    def __trailing_zero(self, num):
        if num == 0:
            return 32 # Assumes 32 bit integer inputs!
        p = 0
        while (num >> p) & 1 == 0:
            p += 1
        return p
    
    def __first_zero(self, num):
        ba = BitArray(uint=num, length=self.__bitarray_size)
        ba.reverse()
        
        oi = 0
        for bit in ba:
            if bit == False:
                break
            oi += 1

        return oi         
    
    def give_estimate(self):
        fzi = self.__first_zero(self.__bitarray.int)
        return (2 ** fzi) / 0.77351
        
    def process(self, element):
        sf = self.__hash(element, self.__hashp) % 2 ** (self.__bitarray_size - 1)
        self.__bitarray.set(1, self.__bitarray_size - 1 - self.__trailing_zero(sf))

We use the **MurmurHash3** function defined in *hash_function* above again when instantiating the class:

In [191]:
fm = flajolet_martin(hash_function)

The Shakespeare-text was processed again, this time without the removal of unique values:

`tr -s  '[:blank:]\n' '\n' < shakespeare.txt | tr 'A-Z' 'a-z' | tr -cd [:lower:][:cntrl:] | sed '/^$/d' > shakespeare_all_words.txt`

And read into a list:

Then, all of the words in *shakespeare_list_all* get processed by the filter:

In [192]:
shakespeare_list_all = file_strip_dict('shakespeare_all_words.txt')

In [193]:
for word in shakespeare_list_all:
    fm.process(word)

In [194]:
print fm.give_estimate()

2647.67100619


The result is *2647.67100619*, however, sort & uniq command gives back *3280* as the exact result - thus the value gotten using the *Flajolet-Martin* algorhtym is too large - it is a:

In [195]:
abs((fm.give_estimate() - 3280) / 3280) * 100

19.278322981934433

19.28% error.

To reduce the error, a common solution is to use the median which is less prone to be influenced by outliers. The problem with this is that the results can only take the form $2^{R} / 0.77351$, where *R* is integer. A common solution is to combine both the mean and the median: Create $k*l$ hash-functions and split them into *k* distinct groups (each of size *l*). Within each group use the median for aggregating together the *l* results, and finally take the mean of the *k* group estimates as the final estimate.

Thus, a new hashing function is made to be passed to our *Flajolet-Martin* class; which uses $k*l$ hash-functions as described above, for each pass of hashing of the same data again:

In [197]:
import random
import numpy

k = 10
l = 10

# first, generating k*l different random hash functions
# which will be having a table of different random seeds for mmh3
hashgroups = []
for k_i in range(k):
    hashseeds = []
    for l_i in range(l):
        hashseeds.append(random.randint(-2**31, 2**31))
    hashgroups.append(hashseeds)

ks = []
for k in hashgroups:
    ls = []
    for l in k:
        # Creating a FM class with this seed.
        fm_kl = flajolet_martin(hash_function, l)
        
        for word in shakespeare_list_all:
            fm_kl.process(word)

        ls.append(fm_kl.give_estimate())
        del fm_kl
    ks.append(numpy.median(ls))

print numpy.mean(ks)

3243.39698259


The result is *3243.39698259* then, and the error is:

In [198]:
abs((numpy.mean(ks) - 3280) / 3280) * 100

1.1159456528696825

Around 1% - depending on the seed values.

### Conclusion
Finally, we can conclude, that the *Flajolet-Martin* algorhytm is a memory-efficient way to approximate the number of unique items in a stream. However, in its basic form (examined in the former part of this section), the values can be off - meaning errors well over 10% - due to some kind-of non-uniformness of the hash-functions, producing an off-value. Should be noted too, that since the approximation is $2^{R} / 0.77351$, the values are somewhat quantized to the powers of 2 too.

In the latter part, the calculation was done using the "median-trick" again, by doing *k* groups having *l* different hash functions over the input, and taking the average of the median of the *l* functions. This greatly reduced the error, however of course it was much more CPU-intensive.